### PR TITLE
feat: Create magic_number.cpp

### PR DESCRIPTION
**Pull request by neha-hasija17**
This Pull Request is for HacktoberFest 2020

#### Checklist

- [ ] Added description of change
- [ ] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [ ] Added tests and example, test must pass
- [ ] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [ ] Relevant documentation/comments is changed or added
- [ ] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [ ] Search previous suggestions before making a new one, as yours may be a duplicate.
- [ ] I acknowledge that all my contributions will be made under the project's license.

Notes: 

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1318"></a>

### CODE REVIEW

File: math/magic_number.cpp [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    int result=n%9;
+    //if result is 1 then the number is a magic number else not
+    if(result==1) return true;
+    else return false;
+}
+
+/** Test function
+ * @returns void
+ */
+void tests() {
```
Comment (Panquesito7) [inline math/magic_number.cpp:39]:
```suggestion
static void tests() {
```

File: math/magic_number.cpp [hunk 2]
```diff
@@ -0,0 +1,67 @@
+/**
+ * @file magic_number.cpp
```
Comment (Panquesito7) [inline math/magic_number.cpp:2]:
```suggestion
 * @file
```

File: math/magic_number.cpp [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+ * @brief A simple program to check if the given number is a magic number or not.
+ * A number is said to be a magic number, if the sum of its digits are calculated
```
Comment (Panquesito7) [inline math/magic_number.cpp:4]:
```suggestion
 *
 * @details
 * A number is said to be a magic number, if the sum of its digits are calculated 
```

File: math/magic_number.cpp [hunk 4]
```diff
@@ -0,0 +1,67 @@
+/**
+ * @file magic_number.cpp
+ * @brief A simple program to check if the given number is a magic number or not.
+ * A number is said to be a magic number, if the sum of its digits are calculated 
+ * till a single digit recursively by adding the sum of the digits after every addition. 
+ * If the single digit comes out to be 1,then the number is a magic number.
+ * @detail This is a shortcut method to verify Magic Number.
```
Comment (Panquesito7) [inline math/magic_number.cpp:7]:
```suggestion
 *
 * This is a shortcut method to verify Magic Number. 
```

File: math/magic_number.cpp [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+ */
+#include <cassert>
+#include <iostream>
+
+/**
+ * Function to check if the given number is magic number or not.
+ * @param n number to be checked.
+ * @return if number is a magic number, returns true, else false.
+ */
+
```
Comment (Panquesito7) [inline math/magic_number.cpp:24]:
```suggestion
```

File: math/magic_number.cpp [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+ * are also divisible by 9. Therefore, if a number is divisible by 9, then, recursively, 
+ * all the digit sums are also divisible by 9. The final digit sum is always 9. 
+ * An increase of 1 in the original number will increase the ultimate value by 1, 
+ * making it 10 and the ultimate sum will be 1, thus verifying that it is a magic number.
+ * @author [Neha Hasija](https://github.com/neha-hasija17)
+ */
+#include <cassert>
+#include <iostream>
+
+/**
```
Comment (Panquesito7) [inline math/magic_number.cpp:19]:
```suggestion
/**
 * @namespace math
 * @brief Math algorithms
 */
namespace math {
/**
```

File: math/magic_number.cpp [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+bool magic_number(int n) {
+    if (n <= 0) {
+        return false;
+    }
+    // result stores the modulus of @param n with 9
+    int result=n%9;
+    //if result is 1 then the number is a magic number else not
+    if(result==1) return true;
+    else return false;
+}
```
Comment (Panquesito7) [inline math/magic_number.cpp:34]:
```suggestion
}
}  // namespace math
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
👍 Well-written and documented code.
Please enable [GitHub Actions](https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions) in your repository of this fork in this link: https://github.com/neha-hasija17/C-Plus-Plus/actions

**Comment by neha-hasija17** [comment]
> 👍 Well-written and documented code.
> Please enable [GitHub Actions](https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions) in your repository of this fork in this link: https://github.com/neha-hasija17/C-Plus-Plus/actions

thankyou sir. please review this.

**Comment by neha-hasija17** [comment]
@Panquesito7 please review this. this Pull request is for hacktober fest 2020

File: math/magic_number.cpp [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+ * number else not. The divisibility rule of 9 says that a number is divisible
+ * by 9 if the sum of its digits are also divisible by 9. Therefore, if a number
+ * is divisible by 9, then, recursively, all the digit sums are also divisible
+ * by 9. The final digit sum is always 9. An increase of 1 in the original
+ * number will increase the ultimate value by 1, making it 10 and the ultimate
+ * sum will be 1, thus verifying that it is a magic number.
+ * @author [Neha Hasija](https://github.com/neha-hasija17)
+ */
+#include <cassert>
+#include <iostream>
```
Comment (Panquesito7) [inline math/magic_number.cpp:20]:
```suggestion
#include <cassert>  /// for assert
#include <iostream> /// for io operations
```

File: math/magic_number.cpp [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    int result = n % 9;
+    // if result is 1 then the number is a magic number else not
+    if (result == 1) {
+        return true;
+    } else {
+        return false;
+    }
+}
+
+/** Test function
```
Comment (Panquesito7) [inline math/magic_number.cpp:41]:
```suggestion
/**
 * @brief Test function
```

File: math/magic_number.cpp [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+    std::cout << "Test 5:\t n=-35\n";
+    assert(magic_number(-35) == false);
+    std::cout << "passed\n";
+}
+
+/** Main function
+ * @returns 0 on exit
+ */
+int main() {
+    tests();
```
Comment (Panquesito7) [inline math/magic_number.cpp:70]:
```suggestion
    tests(); // execute the tests
```

File: math/magic_number.cpp [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+    std::cout << "Test 4:\t n=479001600\n";
+    assert(magic_number(479001600) == false);
+    std::cout << "passed\n";
+
+    std::cout << "Test 5:\t n=-35\n";
+    assert(magic_number(-35) == false);
+    std::cout << "passed\n";
+}
+
+/** Main function
```
Comment (Panquesito7) [inline math/magic_number.cpp:66]:
```suggestion
/**
 * @brief Main function
```

File: math/magic_number.cpp [hunk 12]
```diff
@@ -0,0 +1,72 @@
+/**
+ * @file
+ * @brief A simple program to check if the given number is a magic number or
+ * not. A number is said to be a magic number, if the sum of its digits are
```
Comment (Panquesito7) [inline math/magic_number.cpp:4]:
```suggestion
 * not.
 *
 * @details
 * A number is said to be a magic number, if the sum of its digits are
```

File: math/magic_number.cpp [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+ * is a magic number.
+ *
+ * This is a shortcut method to verify Magic Number.
+ * On dividing the input by 9, if the remainder is 1 then the number is a magic
+ * number else not. The divisibility rule of 9 says that a number is divisible
+ * by 9 if the sum of its digits are also divisible by 9. Therefore, if a number
+ * is divisible by 9, then, recursively, all the digit sums are also divisible
+ * by 9. The final digit sum is always 9. An increase of 1 in the original
+ * number will increase the ultimate value by 1, making it 10 and the ultimate
+ * sum will be 1, thus verifying that it is a magic number.
```
Comment (Panquesito7) [inline math/magic_number.cpp:16]:
```suggestion
 * sum will be 1, thus verifying that it is a magic number.
 *
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Please fix [`clang-tidy`](https://github.com/TheAlgorithms/C-Plus-Plus/pull/1318/checks?check_run_id=1272084106) warnings.

File: math/magic_number.cpp (hunk 7 above)
Comment (Panquesito7) [reply math/magic_number.cpp:39]: Not resolved yet (same as above; this is **necessary**).

**Comment by neha-hasija17** [comment]
@Panquesito7 why the basic tests are failing? I don't understand this, kindly make the necessary changes so it works well like before.

File: math/magic_number.cpp (hunk 6 above)
Comment (Panquesito7) [reply math/magic_number.cpp:22]: Not resolved yet.

File: math/magic_number.cpp [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+}
+}  // namespace math
+
+/**
+ * @brief Test function
+ * @returns void
+ */
+static void tests() {
+    std::cout << "Test 1:\t n=60\n";
+    assert(magic_number(60) == false);
```
Comment (Panquesito7) [inline math/magic_number.cpp:53]:
```suggestion
    assert(math::magic_number(60) == false);
```

File: math/magic_number.cpp [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+    std::cout << "passed\n";
+
+    std::cout << "Test 2:\t n=730\n";
+    assert(magic_number(730) == true);
```
Comment (Panquesito7) [inline math/magic_number.cpp:57]:
```suggestion
    assert(math::magic_number(730) == true);
```

File: math/magic_number.cpp [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+    std::cout << "passed\n";
+
+    std::cout << "Test 3:\t n=0\n";
+    assert(magic_number(0) == false);
```
Comment (Panquesito7) [inline math/magic_number.cpp:61]:
```suggestion
    assert(math::magic_number(0) == false);
```

File: math/magic_number.cpp [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+    std::cout << "passed\n";
+
+    std::cout << "Test 4:\t n=479001600\n";
+    assert(magic_number(479001600) == false);
```
Comment (Panquesito7) [inline math/magic_number.cpp:65]:
```suggestion
    assert(math::magic_number(479001600) == false);
```

File: math/magic_number.cpp [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+    std::cout << "passed\n";
+
+    std::cout << "Test 5:\t n=-35\n";
+    assert(magic_number(-35) == false);
```
Comment (Panquesito7) [inline math/magic_number.cpp:69]:
```suggestion
    assert(math::magic_number(-35) == false);
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

**Comment by neha-hasija17** [comment]
@Panquesito7 please have a look

File: math/magic_number.cpp [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+    std::cout << "Test 4:\t n=479001600\n";
+    assert(math::magic_number(479001600) == false);
+    std::cout << "passed\n";
+
+    std::cout << "Test 5:\t n=-35\n";
+    assert(math::magic_number(-35) == false);
+    std::cout << "passed\n";
+}
+
+/** Main function
```
Comment (Panquesito7) [inline math/magic_number.cpp:73]:
```suggestion
/**
 * @brief Main function
```

**Comment by neha-hasija17** [comment]
@Panquesito7 Kindly review this.

File: math/magic_number.cpp [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+ * @namespace math
+ * @brief Mathematical algorithms
+ */
+namespace math {
+/**
+ * Function to check if the given number is magic number or not.
+ * @param n number to be checked.
+ * @return if number is a magic number, returns true, else false.
+ */
+bool magic_number(int n) {
```
Comment (Panquesito7) [inline math/magic_number.cpp:32]:
```suggestion
bool magic_number(const uint64_t &n) {
```

File: math/magic_number.cpp [hunk 21]
```diff
@@ -0,0 +1,80 @@
...
+    if (n <= 0) {
+        return false;
+    }
+    // result stores the modulus of @param n with 9
+    int result = n % 9;
```
Comment (Panquesito7) [inline math/magic_number.cpp:37]:
```suggestion
    uint64_t result = n % 9;
```

**Review by Panquesito7 — APPROVED** [review]
Neat, code looks pretty well; LGTM. 😄 👍

**Review by ayaankhan98 — APPROVED** [review]
LGTM! 
Thanks for your contribution

**Comment by Panquesito7** [comment]
@ayaankhan98 Please use squash-merge. Thank you.

**Pull request merged**